File: csharp/uwapi-cs/commands.cs
using System.Runtime.InteropServices;

namespace Unnatural
{
    using Order = Interop.UwOrder;

    public static class Commands
    {
        public static Order[] Orders(uint unit)
        {
            Interop.UwOrders os = new Interop.UwOrders();
            Interop.uwOrders(unit, ref os);
            Order[] tmp = new Order[os.count];
            for (int i = 0; i < os.count; i++)
                tmp[i] = Marshal.PtrToStructure<Order>(os.orders + i * Marshal.SizeOf<Order>());
            return tmp;
        }

        public static void Order(uint unit, Order order)
        {
            Interop.uwOrder(unit, ref order);
        }

        public static Order Stop()
        {
            Order o = new Order();
            o.entity = 4294967295;
            o.position = 4294967295;
            o.order = 1;
            o.priority = 2;
            return o;
        }

        public static Order Guard()
        {
            Order o = new Order();
            o.entity = 4294967295;
            o.position = 4294967295;
            o.order = 2;
            o.priority = 2;
            return o;
        }

        public static Order RunToPosition(uint position)
        {
            Order o = new Order();
            o.entity = 4294967295;
            o.position = position;
            o.order = 3;
            o.priority = 2;
            return o;
        }

        public static Order RunToEntity(uint entity)
        {
            Order o = new Order();
            o.entity = entity;
            o.position = 4294967295;
            o.order = 3;
            o.priority = 2;
            return o;
        }

        public static Order FightToPosition(uint position)
        {
            Order o = new Order();
            o.entity = 4294967295;
            o.position = position;
            o.order = 4;
            o.priority = 2;
            return o;
        }

        public static Order FightToEntity(uint entity)
        {
            Order o = new Order();
            o.entity = entity;
            o.position = 4294967295;
            o.order = 4;
            o.priority = 2;
            return o;
        }

        public static void CommandSelfDestruct(uint unit)
        {
            Interop.uwCommandSelfDestruct(unit);
        }

        public static void CommandPlaceConstruction(uint proto, uint position, float yaw = 0)
        {
            Interop.uwCommandPlaceConstruction(proto, position, yaw);
        }

        public static void CommandSetRecipe(uint unit, uint recipe)
        {
            Interop.uwCommandSetRecipe(unit, recipe);
        }

        public static void CommandLoad(uint unit, uint resourceType)
        {
            Interop.uwCommandLoad(unit, resourceType);
        }

        public static void CommandUnload(uint unit)
        {
            Interop.uwCommandUnload(unit);
        }

        public static void CommandMove(uint unit, uint position, float yaw = 0)
        {
            Interop.uwCommandMove(unit, position, yaw);
        }

        public static void CommandAim(uint unit, uint target)
        {
            Interop.uwCommandAim(unit, target);
        }

        public static void CommandRenounceControl(uint unit)
        {
            Interop.uwCommandRenounceControl(unit);
        }
    }
}

File: python/uw/uw/commands.py
from .helpers import Order
from .helpers import OrderType
from .helpers import OrderPriority


class Commands:
    invalid = 4294967295

    def __init__(self, api, ffi):
        self._api = api
        self._ffi = ffi

    def orders(self, unit: int) -> list[Order]:
        os = self._ffi.new("struct UwOrders *")
        self._api.uwOrders(unit, os)
        return [Order.from_c(o) for o in self._ffi.unpack(os.orders, os.count)]

    def order(self, unit: int, order: Order):
        self._api.uwOrder(unit, order)

    def stop(self) -> Order:
        return Order(entity=self.invalid, position=self.invalid, order_type=OrderType.Stop, priority=OrderPriority.User)

    def guard(self) -> Order:
        return Order(entity=self.invalid, position=self.invalid, order_type=OrderType.Guard,
                     priority=OrderPriority.User)

    def run_to_position(self, position: int) -> Order:
        return Order(entity=self.invalid, position=position, order_type=OrderType.Run,
                     priority=OrderPriority.User)

    def run_to_entity(self, entity: int) -> Order:
        return Order(entity=entity, position=self.invalid, order_type=OrderType.Run,
                     priority=OrderPriority.User)

    def fight_to_position(self, position: int) -> Order:
        return Order(entity=self.invalid, position=position, order_type=OrderType.Fight,
                     priority=OrderPriority.User)

    def fight_to_entity(self, entity: int) -> Order:
        return Order(entity=entity, position=self.invalid, order_type=OrderType.Fight,
                     priority=OrderPriority.User)

    def command_self_destruct(self, unit: int):
        self._api.uwCommandSelfDestruct(unit)

    def command_place_construction(self, proto: int, position: int, yaw: float = 0):
        self._api.uwCommandPlaceConstruction(proto, position, yaw)

    def command_set_recipe(self, unit: int, recipe: int):
        self._api.uwCommandSetRecipe(unit, recipe)

    def command_load(self, unit: int, resource_type: int):
        self._api.uwCommandLoad(unit, resource_type)

    def command_unload(self, unit: int):
        self._api.uwCommandUnload(unit)

    def command_move(self, unit: int, position: int, yaw: float = 0):
        self._api.uwCommandMove(unit, position, yaw)

    def command_aim(self, unit: int, target: int):
        self._api.uwCommandAim(unit, target)

    def command_renounce_control(self, unit: int):
        self._api.uwCommandRenounceControl(unit)
